File: src/cpp/core/inference_engine.cpp

#include "inference_engine.hpp"
#include <iostream>
#include <algorithm>

namespace yolov11 {

InferenceEngine::InferenceEngine() {
    // Constructor
}

InferenceEngine::~InferenceEngine() {
    // Destructor
}

void InferenceEngine::load_model(const std::string& model_path, const std::string& device_name) {
    try {
        std::cout << "Loading model: " << model_path << " on " << device_name << std::endl;
        
        // 1. Read Model
        model = core.read_model(model_path);
        
        // 2. Preprocessing (Basic setup via OpenVINO PPP could be done here, but we'll do manual for clarity/control)
        // Check input/output shapes
        if (model->inputs().size() != 1) {
            throw std::runtime_error("Model must have exactly one input.");
        }
        
        input_name = model->input().get_any_name();
        input_shape = model->input().get_shape();
        
        // Handle dynamic shapes if necessary (assuming fixed for standard YOLO export, e.g. [1,3,640,640])
        // If dynamic, we might need to reshape or stick to a target size.
        
        // 3. Compile Model
        compiled_model = core.compile_model(model, device_name);
        infer_request = compiled_model.create_infer_request();
        
        std::cout << "Model loaded successfully." << std::endl;
        std::cout << "Input Shape: " << input_shape << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "Error loading model: " << e.what() << std::endl;
        throw;
    }
}

cv::Mat InferenceEngine::preprocess(const cv::Mat& frame) {
    // 1. Resize to model input size (e.g., 640x640)
    // Assuming input_shape is [1, 3, H, W]
    int w = input_shape[3];
    int h = input_shape[2];
    
    cv::Mat resized;
    // Use letterbox if strict aspect ratio is needed, but simple resize for MVP if ok.
    // YOLOv11 usually expects letterbox (pad to square) for best accuracy.
    // Let's implement basic resize for now, improve to letterbox if requested.
    cv::resize(frame, resized, cv::Size(w, h));
    
    // 2. Convert Color (BGR -> RGB)
    cv::cvtColor(resized, resized, cv::COLOR_BGR2RGB);
    
    return resized;
}

std::vector<Detection> InferenceEngine::infer(const cv::Mat& frame, float conf_threshold, float iou_threshold) {
    // Basic Timing
    auto start = std::chrono::high_resolution_clock::now();

    // 1. Preprocess
    cv::Mat resized_img = preprocess(frame);
    
    // 2. Set Input Tensor
    // OpenVINO expects NCHW, OpenCV is NHWC (HWC actually).
    // We need to permute.
    // Or we can wrap existing Mat if we used PPP.
    
    // Get input tensor from request
    ov::Tensor input_tensor = infer_request.get_input_tensor();
    
    // Copy data and change layout (HWC -> NCHW)
    // shape: [1, 3, 640, 640]
    float* input_data = input_tensor.data<float>();
    
    int H = input_shape[2];
    int W = input_shape[3];
    int C = input_shape[1];
    
    // Normalize 0-255 -> 0.0-1.0
    for (int h = 0; h < H; h++) {
        for (int w = 0; w < W; w++) {
            cv::Vec3b pixel = resized_img.at<cv::Vec3b>(h, w);
            input_data[0*H*W + h*W + w] = pixel[0] / 255.0f; // R
            input_data[1*H*W + h*W + w] = pixel[1] / 255.0f; // G
            input_data[2*H*W + h*W + w] = pixel[2] / 255.0f; // B
        }
    }
    
    // 3. Infer
    infer_request.infer();
    
    // 4. Postprocess
    const ov::Tensor& output_tensor = infer_request.get_output_tensor();
    const float* output_buffer = output_tensor.data<const float>();
    
    // Output shape for YOLO usually [1, 84, 8400] (for 80 classes + 4 bbox coords)
    // 84 = 4 coord + 80 classes
    // We need to verify the output shape handling.
    
    auto end = std::chrono::high_resolution_clock::now();
    // std::cout << "Inference time: " << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << "ms" << std::endl;

    return postprocess(output_buffer, frame.size(), conf_threshold, iou_threshold);
}

std::vector<Detection> InferenceEngine::postprocess(const float* output_data, const cv::Size& original_size, float conf_threshold, float iou_threshold) {
    std::vector<Detection> detections;
    std::vector<int> class_ids;
    std::vector<float> confidences;
    std::vector<cv::Rect> boxes;
    
    // YOLOv11 / v8 Output Format: [Batch, Channels, Anchors] -> [1, 4 + NumClasses, NumProposals]
    // e.g. [1, 84, 8400]
    
    // We assume the shape is available.
    ov::Shape out_shape = infer_request.get_output_tensor().get_shape();
    int num_classes = out_shape[1] - 4; // e.g. 84 - 4 = 80
    int num_proposals = out_shape[2];   // e.g. 8400
    
    // The data is likely row-major but the shape implies distinct channels per proposal.
    // Actually YOLOv8/11 export logic suggests [1, 84, 8400]
    // where 84 is (cx, cy, w, h, class0_conf, class1_conf...)
    
    // We transpose conceptually by iterating:
    // For each proposal p (0..8400):
    //   cx = data[0 * num_proposals + p]
    //   cy = data[1 * num_proposals + p]
    //   ...
    
    const float* px = output_data;
    const float* py = px + num_proposals;
    const float* pw = py + num_proposals;
    const float* ph = pw + num_proposals;
    const float* p_classes = ph + num_proposals; // Start of class scores
    
    float ratio_w = (float)original_size.width / input_shape[3];
    float ratio_h = (float)original_size.height / input_shape[2];
    
    for (int i = 0; i < num_proposals; i++) {
        
        // Find best class score for this proposal
        float max_score = 0.0f;
        int max_class_id = -1;
        
        for (int c = 0; c < num_classes; c++) {
            float score = p_classes[c * num_proposals + i];
            if (score > max_score) {
                max_score = score;
                max_class_id = c;
            }
        }
        
        if (max_score > conf_threshold) {
            float cx = px[i];
            float cy = py[i];
            float w = pw[i];
            float h = ph[i];
            
            // Convert to top-left x,y
            int left = static_cast<int>((cx - 0.5 * w) * ratio_w);
            int top = static_cast<int>((cy - 0.5 * h) * ratio_h);
            int width = static_cast<int>(w * ratio_w);
            int height = static_cast<int>(h * ratio_h);
            
            boxes.push_back(cv::Rect(left, top, width, height));
            class_ids.push_back(max_class_id);
            confidences.push_back(max_score);
        }
    }
    
    // NMS
    std::vector<int> indices;
    cv::dnn::NMSBoxes(boxes, confidences, conf_threshold, iou_threshold, indices);
    
    for (int idx : indices) {
        Detection det;
        det.class_id = class_ids[idx];
        det.confidence = confidences[idx];
        det.box = boxes[idx];
        det.class_name = std::to_string(class_ids[idx]); // Map to name later
        detections.push_back(det);
    }
    
    return detections;
}

} // namespace yolov11

File: src/rust/src/core/inference_engine.rs
use anyhow::{Context, Result};
use opencv::{
    core::{Mat, Point, Rect, Scalar, Size, Vector},
    dnn, imgproc,
    prelude::*,
};
use openvino::{Core, ElementType, Tensor};
use std::sync::Arc;

pub struct Detection {
    pub class_id: i32,
    pub score: f32,
    pub box_rect: Rect,
    pub class_name: String,
}

pub struct InferenceEngine {
    core: Core,
    compiled_model: Option<openvino::CompiledModel>,
    infer_request: Option<openvino::InferRequest>,
    input_width: usize,
    input_height: usize,
}

impl InferenceEngine {
    pub fn new() -> Self {
        let core = Core::new().expect("Failed to create OpenVINO Core");
        Self {
            core,
            compiled_model: None,
            infer_request: None,
            input_width: 640,
            input_height: 640,
        }
    }

    pub fn load_model(&mut self, model_path: &str, device: &str) -> Result<()> {
        println!(
            "Loading model (OpenVINO 0.7.1 - NMS Enabled): {} on {}",
            model_path, device
        );
        let bin_path = model_path.replace(".xml", ".bin");

        let model = self.core.read_model_from_file(model_path, &bin_path)?;
        let mut compiled_model = self.core.compile_model(&model, device.into())?;
        let infer_request = compiled_model.create_infer_request()?;

        self.compiled_model = Some(compiled_model);
        self.infer_request = Some(infer_request);

        println!("Model loaded successfully");
        Ok(())
    }

    pub fn infer(&mut self, frame: &Mat, conf_threshold: f32) -> Result<Vec<Detection>> {
        let request = self
            .infer_request
            .as_mut()
            .ok_or(anyhow::anyhow!("Model not loaded"))?;

        // 1. Preprocess
        let img = letterbox(frame, self.input_height as i32, self.input_width as i32)?;
        // OpenCV Rust Mat is already usable, but need to extract data.

        // Convert to RGB
        let mut rgb = Mat::default();
        imgproc::cvt_color(&img, &mut rgb, imgproc::COLOR_BGR2RGB, 0)?;

        // Convert to Float32 and Normalize 0-1
        let mut float_img = Mat::default();
        rgb.convert_to(&mut float_img, opencv::core::CV_32F, 1.0 / 255.0, 0.0)?;

        // HWC to CHW
        let h = self.input_height;
        let w = self.input_width;
        let mut input_data = vec![0.0f32; 3 * h * w];

        for y in 0..h {
            for x in 0..w {
                let pixel = float_img.at_2d::<opencv::core::Vec3f>(y as i32, x as i32)?;
                input_data[0 * h * w + y * w + x] = pixel[0]; // R
                input_data[1 * h * w + y * w + x] = pixel[1]; // G
                input_data[2 * h * w + y * w + x] = pixel[2]; // B
            }
        }

        // 2. Set Input
        let mut input_tensor = request.get_input_tensor()?;
        // Note: 0.7.1 might return Tensor, which has get_data_mut (generic?)
        // The reference code used: input_tensor.get_data_mut::<f32>()?
        {
            let buffer = input_tensor.get_data_mut::<f32>()?;
            if buffer.len() != input_data.len() {
                return Err(anyhow::anyhow!(
                    "Input tensor size mismatch: expected {}, got {}",
                    buffer.len(),
                    input_data.len()
                ));
            }
            buffer.copy_from_slice(&input_data);
        }

        // 3. Infer
        request.infer()?;

        // 4. Post-process (YOLOv11 NMS)
        let output_tensor = request.get_output_tensor()?;
        let output_data = output_tensor.get_data::<f32>()?;
        let len = output_data.len();

        let mut anchors = 8400;
        let mut channels = len / anchors;

        if len % anchors != 0 {
            anchors = 8400;
            channels = 84;
        }

        let mut boxes: Vector<Rect> = Vector::new();
        let mut scores: Vector<f32> = Vector::new();
        let mut class_ids: Vec<i32> = Vec::new();

        let scale_x = frame.cols() as f32 / w as f32;
        let scale_y = frame.rows() as f32 / h as f32;

        for i in 0..anchors {
            let mut max_score = 0.0f32;
            let mut cls_id = 0;

            // Classes start after the first 4 bbox coordinates
            for cls in 0..(channels - 4) {
                let idx = (4 + cls) * anchors + i;
                if idx < len {
                    let val = output_data[idx];
                    if val > max_score {
                        max_score = val;
                        cls_id = cls;
                    }
                }
            }

            if max_score > conf_threshold {
                let cx = output_data[0 * anchors + i];
                let cy = output_data[1 * anchors + i];
                let bw = output_data[2 * anchors + i];
                let bh = output_data[3 * anchors + i];

                let x = ((cx - bw / 2.0) * scale_x) as i32;
                let y = ((cy - bh / 2.0) * scale_y) as i32;
                let w_box = (bw * scale_x) as i32;
                let h_box = (bh * scale_y) as i32;

                boxes.push(Rect::new(x, y, w_box, h_box));
                scores.push(max_score);
                class_ids.push(cls_id as i32);
            }
        }

        // NMS
        let mut indices: Vector<i32> = Vector::new();
        dnn::nms_boxes(&boxes, &scores, conf_threshold, 0.45, &mut indices, 1.0, 0)?;

        let mut detections = Vec::new();
        for i in 0..indices.len() {
            let idx = indices.get(i)? as usize;
            let box_rect = boxes.get(idx)?;
            let score = scores.get(idx)?;
            let class_id = class_ids[idx];

            // Map class_id to name
            let class_name = match class_id {
                0 => "fire",
                1 => "smoke",
                2 => "fire_smoke",
                3 => "steam",
                _ => "unknown",
            }
            .to_string();

            detections.push(Detection {
                class_id,
                score,
                box_rect,
                class_name,
            });
        }

        Ok(detections)
    }
}

// Helper
fn letterbox(img: &Mat, new_h: i32, new_w: i32) -> Result<Mat> {
    let h = img.rows();
    let w = img.cols();
    let r = (new_h as f32 / h as f32).min(new_w as f32 / w as f32);
    let new_unpad_w = (w as f32 * r).round() as i32;
    let new_unpad_h = (h as f32 * r).round() as i32;
    let dw = (new_w - new_unpad_w) / 2;
    let dh = (new_h - new_unpad_h) / 2;

    let mut resized = Mat::default();
    imgproc::resize(
        img,
        &mut resized,
        Size::new(new_unpad_w, new_unpad_h),
        0.0,
        0.0,
        imgproc::INTER_LINEAR,
    )?;

    let mut result = Mat::default();
    opencv::core::copy_make_border(
        &resized,
        &mut result,
        dh,
        dh,
        dw,
        dw,
        opencv::core::BORDER_CONSTANT,
        Scalar::new(114.0, 114.0, 114.0, 0.0),
    )?;

    Ok(result)
}
